packages/react/types/jsx-namespace.d.ts

```diff
@@ -2,8 +2,12 @@ import 'react'
 import { Interpolation } from '@emotion/serialize'
 import { Theme } from './index'
 
-type WithConditionalCSSProp<P> = 'className' extends keyof P
-  ? (P extends { className?: string } ? P & { css?: Interpolation<Theme> } : P)
+// prettier-ignore
+type WithConditionalCSSProp<P> =
+  'className' extends keyof P
+    ? string extends P['className' & keyof P]
```

Comment:
Could u explain the machinery if this type step by step? Especially the `& keyof` bit - is this just supposed to produce never if there is no className prop?

Ill have to think the whole thing through in terms of possible perf regressions as well.

Reply:
sure. I assume you get the gists of it, so I'll focus on that confusing `& keyof`.

Let's just translate it into pseudocode, so we're on the same page.
```
  if className is a key in Props
    then if Props["className"] can be a string
      then return Props merged with { css?: Interpolation<Theme> }
      else return Props
    else return Props
```

Previous code (pasted below for comparison) was checking _can I assign Props to `{ className?: string }`_.
You can assign `{ className?: undefined }` to `{ className?: string }`, and you can't assign `{ className: string[] | string }` or `{ className: unknown }` to it.

```
type WithConditionalCSSProp<P> = 'className' extends keyof P
  ? (P extends { className?: string } ? P & { css?: Interpolation<Theme> } : P)
  : P
```

---

> Especially the & keyof bit - is this just supposed to produce never if there is no className prop?

Yup, that's correct.

![image](https://user-images.githubusercontent.com/15332326/100552740-4b4a4900-3289-11eb-92a5-8559d0a2c53c.png)

This is the most "interesting" part. The props will surely be an object, and we just care about the `className` prop.
But we can't read that `className` from `P` until we know that such a key exists in `P`. 

![image](https://user-images.githubusercontent.com/15332326/100552808-b5fb8480-3289-11eb-9d2e-faf8450002b3.png)

This code is sufficient in TS 4.0, but TS 3.6 is a bit dumber.
I had to add that `& keyof P`, because `className` is really a key of `P` at this point, and we'll never get `never` there. TS 3.6 just doesn't remember that we already checked it :)

But hey! Doesn't this mean that we can get rid of that first `extends keyof P` clause and use only one conditional type?

Like so
```tsx
type WithConditionalCSSProp<P> =
  string extends P['className' & keyof P]
    ? P & { css?: Interpolation<Theme> }
    : P
```

We could if we didn't care about TS 3.4 anymore.

![image](https://user-images.githubusercontent.com/15332326/100552904-91ec7300-328a-11eb-9a13-038d9249931b.png)

3.4 is before TypeScript was smart enough to know that you can't be two different strings at once. An intersection of `'any-string' & 'className'` doesn't collapse into `never`, we get an error, and a nice `any`, `css` is added to _all props_ regardless if they have className and our tests fail.

![image](https://user-images.githubusercontent.com/15332326/100553137-4cc94080-328c-11eb-8b5d-8d92e9d25f20.png)

That `dtslint` was a smart choice.

> Ill have to think the whole thing through in terms of possible perf regressions as well.

Do you have any benchmarks / performance tests for Emotion types?

Reply:
I have done Very Scientific Research™ by checking a fresh CRA with just a few css prop-based components and those are some numbers from the `--diagnostics` (I've ignored mem consumption and times as those are not stable between reruns):

### Current

```
Nodes:           336996
Identifiers:     114711
Symbols:          81060
Types:             4642
Instantiations:   14645
```

### This PR

```
Nodes:           336998
Identifiers:     114711
Symbols:          80811
Types:             4181
Instantiations:    4824
```

### Without `& keyof P`

```
Nodes:           336994
Identifiers:     114710
Symbols:          80811
Types:             4180
Instantiations:    4809
```

--- 

As we may notice there is only a rather negligible difference between your PR and the "without `& keyof P`" variant that aims to satisfy older TS versions **but** there is actually **a big improvement** over the current version in terms of instantiations that are often responsible for slowdowns